src/Libraries/Microsoft.Extensions.AI/ChatCompletion/StreamingChatCompletionUpdateExtensions.cs

```diff
@@ -0,0 +1,266 @@
+    /// <param name="coalesceContent">
+    /// <see langword="true"/> to attempt to coalesce contiguous streaming update <see cref="AIContent"/> items, where applicable,
+    /// into a single <see cref="AIContent"/>, in order to reduce the number of individual content items that are included in
+    /// manufactured <see cref="ChatMessage"/> instances. When <see langword="false"/>, the content items are kept unaltered.
+    /// A value of <see langword="true"/> does not impact the updates yielded from the enumerable, only the final message added
+    /// to <paramref name="chatMessages"/>.
+    /// </param>
+    /// <returns>
+    /// An enumerable that yields the same updates as <paramref name="updates"/>. It must be enumerated to completion
+    /// or disposed to ensure that the final message is added to <paramref name="chatMessages"/>.
+    /// </returns>
+    /// <remarks>
+    /// <para>
+    /// If any updates were yielded, the method will add a message to <paramref name="chatMessages"/> upon
+    /// completion of iteration, regardless of whether completion occurs due to exhausting the input enumerable,
+    /// due to an exception ocurring, or due to early disposal of the enumerator. Each iteration through the
+    /// returned enumerable may possibly add a message, so it should typically only be enumerated once.
+    /// </para>
+    /// <para>
+    /// At most one <see cref="ChatMessage"/> will be added to <paramref name="chatMessages"/>. If the supplied updates
+    /// represent multiple choices, only one choice will be added.
+    /// </para>
+    /// </remarks>
+    public static IAsyncEnumerable<StreamingChatCompletionUpdate> WithMessageAddedAsync(
+        this IAsyncEnumerable<StreamingChatCompletionUpdate> updates,
+        IList<ChatMessage> chatMessages,
+        bool coalesceContent = true)
+    {
+        _ = Throw.IfNull(updates);
+        _ = Throw.IfNull(chatMessages);
+
+        return ImplAsync(updates, chatMessages, coalesceContent);
+
+        static async IAsyncEnumerable<StreamingChatCompletionUpdate> ImplAsync(
+            IAsyncEnumerable<StreamingChatCompletionUpdate> updates,
+            IList<ChatMessage> chatMessages,
+            bool coalesceContent = true)
+        {
+            List<StreamingChatCompletionUpdate> updatesList = [];
+            try
+            {
+                await foreach (var update in updates.ConfigureAwait(false))
+                {
+                    updatesList.Add(update);
+                    yield return update;
+                }
+            }
+            finally
+            {
+                if (coalesceContent)
+                {
+                    CoalesceTextContent(updatesList);
+                }
+
+                if (ToChatCompletion(updatesList) is { Choices.Count: > 0 } completion)
+                {
+                    chatMessages.Add(completion.Message);
```

Comment:
Should we be updating chat history on cancellation as well?

Reply:
Great point. This likely needs to be controllable by a flag.

 * If the app presents interrupted messages as if they cease to exist (i.e., the disappear from the UI), then you wouldn't want to have the message be added if cancelled
 * In the more common case where interrupted messages are left in the UI, you would want the message to be added. Otherwise the AI would behave as if it never answered your previous question even when you can see a partial answer in the UI.

If cancellation is due to user actions, such as submitting a further message or clicking "Stop responding" or similar, it's more likely you do want the partial message to be appended. So it would make sense to default to that behavior.

I'm not sure if there are other common cases for cancellation that aren't triggered by the end user. Well, there are cases like "the application or current window has shut down" but in those cases it's not harmful to append given that we're likely abandoning the message thread anyway.